### PR TITLE
feat: add `ANY/ALL/SOME(array)` operator to support `get_indexes` in sqlalchemy

### PR DESCRIPTION
**Pull request by HuaHuaY**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As the title.
These operators are used in the right side of binary operators which return boolean value. They require that there are some expressions in the parentheses, which return arrays. `SOME` is a synonym for `ANY`.
In PostgreSQL, there are some other operators named `ANY/ALL/SOME(subquery)`. They are not supported in this PR.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### Types of user-facing changes

* SQL commands, functions, and operators

### Release note

support `ANY/ALL/SOME(array)` operator.
PostgreSQL's doc:
ANY/SOME (array): https://www.postgresql.org/docs/current/functions-comparisons.html#id-1.5.8.30.16
ALL (array): https://www.postgresql.org/docs/current/functions-comparisons.html#id-1.5.8.30.17

## Refer to a related PR or issue link (optional)
related #6636

### CODE REVIEW

File: src/sqlparser/src/parser.rs [hunk 1]
```diff
@@ -1046,11 +1046,35 @@ impl Parser {
...
 
         if let Some(op) = regular_binary_operator {
-            Ok(Expr::BinaryOp {
-                left: Box::new(expr),
-                op,
-                right: Box::new(self.parse_subexpr(precedence)?),
-            })
+            if let Some(keyword) =
+                self.parse_one_of_keywords(&[Keyword::ANY, Keyword::ALL, Keyword::SOME])
+            {
```
Comment (BowenXiao1999) [inline src/sqlparser/src/parser.rs:1051]: Add some comments for this? After all most code in sqlparser is not ours.

File: src/sqlparser/src/parser.rs [hunk 2]
```diff
@@ -1046,11 +1046,35 @@ impl Parser {
...
+                self.expect_token(&Token::LParen)?;
+                let right = self.parse_expr()?;
+                self.expect_token(&Token::RParen)?;
```
Comment (BowenXiao1999) [inline src/sqlparser/src/parser.rs:1054]: Can `select 1 < SOME(null);` work?

File: src/frontend/planner_test/tests/testdata/expr.yaml [hunk 3]
```diff
@@ -351,3 +351,98 @@
   batch_plan: |
     BatchProject { exprs: ['integer[][]':Varchar] }
     └─BatchValues { rows: [[]] }
+- sql: |
+    select 1 = SOME(1);
+  binder_error: 'Bind error: op ANY/ALL (array) requires array on right side'
```
[the remark this answers is not in the document]
Comment (HuaHuaY) [reply src/frontend/planner_test/tests/testdata/expr.yaml:356]: This error is copied from PostgreSQL, and there is no `some` in their error output.

File: src/frontend/src/binder/expr/binary_op.rs [hunk 4]
```diff
@@ -43,70 +74,49 @@ impl Binder {
             BinaryOperator::And => ExprType::And,
             BinaryOperator::Or => ExprType::Or,
             BinaryOperator::Like => ExprType::Like,
-            BinaryOperator::NotLike => return self.bind_not_like(bound_left, bound_right),
+            // Err(zehua): `select 'a' not like any(array['b']);`
+            BinaryOperator::NotLike => {
```
[the remark this answers is not in the document]
Comment (HuaHuaY) [reply src/frontend/src/binder/expr/binary_op.rs:77]: I forgot to delete it. It has been solved. Have been fixed.

File: src/frontend/src/binder/expr/binary_op.rs [hunk 5]
```diff
@@ -43,70 +74,49 @@ impl Binder {
...
+            }
             BinaryOperator::BitwiseOr => ExprType::BitwiseOr,
             BinaryOperator::BitwiseAnd => ExprType::BitwiseAnd,
             BinaryOperator::PGBitwiseXor => ExprType::BitwiseXor,
             BinaryOperator::PGBitwiseShiftLeft => ExprType::BitwiseShiftLeft,
             BinaryOperator::PGBitwiseShiftRight => ExprType::BitwiseShiftRight,
-            BinaryOperator::Concat => return self.bind_concat_op(bound_left, bound_right),
-            BinaryOperator::PGRegexMatch => return self.bind_regex_match(bound_left, bound_right),
+            BinaryOperator::Concat => {
+                match (bound_left.return_type(), bound_right.return_type()) {
```
[the remark this answers is not in the document]
Comment (HuaHuaY) [reply src/frontend/src/binder/expr/binary_op.rs:87]: No, `any/all/some` need to refactor `not like`. I chose to refactor the whole `bind_binary_op`.

File: src/sqlparser/src/parser.rs (hunk 2 above)
Comment (BowenXiao1999) [reply src/sqlparser/src/parser.rs:1057]: Sorry I mean `select SOME(null) > 1;`

Comment (HuaHuaY) [reply src/sqlparser/src/parser.rs:1057]: No, we can't. `SOME/ANY/ALL` can only appear in the right side of binary op.

Comment (BowenXiao1999) [reply src/sqlparser/src/parser.rs:1057]: get it

File: src/frontend/src/binder/expr/binary_op.rs [hunk 6]
```diff
@@ -24,11 +24,42 @@ impl Binder {
...
         let bound_right = self.bind_expr(right)?;
-        let func_type = match op {
+
+        func_types.extend(Self::resolve_binary_operator(
+            op,
+            &bound_left,
+            &bound_right,
+        )?);
+
+        FunctionCall::new_binary_op_func(func_types, vec![bound_left, bound_right])
```
[the remark this answers is not in the document]
Comment (HuaHuaY) [reply src/frontend/src/binder/expr/binary_op.rs:53]: I will create a new issue to track this.

File: src/sqlparser/src/parser.rs [hunk 7]
```diff
@@ -1046,11 +1046,40 @@ impl Parser {
...
+                self.expect_token(&Token::LParen)?;
+                // In upstream's PR of parser-rs, there is `self.parser_subexpr(precedence)` here.
+                // But it will fail to parse `select 1 = any(null and true);`.
+                let right = self.parse_expr()?;
+                self.expect_token(&Token::RParen)?;
+
+                // TODO: support `all/any/some(subquery)`.
+                if let Expr::Subquery(_) = &right {
+                    parser_err!("ANY/SOME/ALL(Subquery) is not implemented")?;
+                }
```
[the remark this answers is not in the document]
Comment (HuaHuaY) [reply src/sqlparser/src/parser.rs:1062]: I will create a new issue to implement `all/any/some(subquery)` and avoid inconsistent behavior.

File: src/expr/src/expr/expr_some_all.rs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                    if !visible {
+                        output_array.append_null();
+                        continue;
+                    }
+
+                    output_array.append(
+                        self.compare_left_right(
+                            left.map(|s| s.into_scalar_impl()),
+                            right.map(|s| s.into_scalar_impl()),
+                        )?
```
[the remark this answers is not in the document]
Comment (BowenXiao1999) [reply src/expr/src/expr/expr_some_all.rs:128]: is this what we will change in this PR?

**Review by BowenXiao1999 — COMMENTED** [review]
LGTM for the query part. I'm not so sure about the impl in binder and expression so pls follow @xiangjinwu

File: src/expr/src/expr/expr_some_all.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (HuaHuaY) [reply src/expr/src/expr/expr_some_all.rs:128]: Yes. But I'm doing other work now, so this modification will be delayed for a while.

File: src/expr/src/expr/expr_some_all.rs [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+            capacity,
+        );
+
+        let func_results = self.func.eval(&data_chunk)?;
+        let mut func_results_iter = func_results.as_bool().iter();
+        Ok(Arc::new(
+            num_array
+                .into_iter()
+                .map(|num| {
+                    self.resolve_boolean_vec(func_results_iter.by_ref().take(num).collect_vec())
```
Comment (xiangjinwu) [inline src/expr/src/expr/expr_some_all.rs:160]: When `num` is 0, there are 3 reasons:
* The corresponding row is invisible.
* The rhs array is `null`.
* The rhs array is empty (`array[]`).

Note that PostgreSQL returns different results for a null array and empty array:
```
test=# select 1 < ALL(array[]::int[]);
 ?column? 
----------
 t
(1 row)

test=# select 1 < SOME(array[]::int[]);
 ?column? 
----------
 f
(1 row)

test=# select 1 < ALL(null);
 ?column? 
----------
 
(1 row)

test=# select 1 < SOME(null);
 ?column? 
----------
 
(1 row)
```

**Review by xiangjinwu — COMMENTED** [review]
Most lgtm. Just found more edge cases:

**Review by xiangjinwu — APPROVED** [review]
Fixed remaining issues on his behalf.
